### PR TITLE
[Core] introduce `PeftAdapterMixin` module.

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Delegate the `peft` related methods residing under `ModelMixin` to a separate class called `PeftAdapterMixin` following what's done in `transformers`: https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/peft.py#L43.

It also introduces a new module called `integrations`, again, following `transformers`. I see it playing to our advantage: 

* `modeling_utils.py` becomes cleaner. 
* Clear separation of concerns. 
* If we want to add new integrations (such as with `bitsandbytes`), we'd have a better structure. 

I don't suspect anything introduced in this PR to be backwards-breaking but happy to be proven wrong and fix it.

### CODE REVIEW

**Review by DN6 — COMMENTED** [review]
Makes sense to me. Think we could also do this for xformers.

**Comment by sayakpaul** [comment]
> Think we could also do this for xformers.

`xformers` -- I think not as it's dealt similarly as SDPA, by means of an attention processor.

File: src/diffusers/integrations/__init__.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
Comment (patrickvonplaten) [inline src/diffusers/integrations/__init__.py:1]: Shouldn't we rather add this file under `src/diffusers/loaders/...`? I'm not sure it's a good idea to add `integrations` as a new folder here

**Review by patrickvonplaten — COMMENTED** [review]
Nice idea to factor out PEFT into a Mixin! 
Two questions:
- 1.) It was already the case before, but do we really want to allow PEFT adaptors for **all** models. E.g. does it make sense that a VAE can also add LoRA layers? 
- 2.) Can we maybe move the Mixin to `src/diffusers/loaders` ?

**Comment by sayakpaul** [comment]
> It was already the case before, but do we really want to allow PEFT adaptors for all models. E.g. does it make sense that a VAE can also add LoRA layers?

This is what I think too. So, instead of adding the PEFT mixin to `ModelMixin` we can be selective and only add to the models that we know are used in LoRA training (from our official scripts). WDYT? 

> Can we maybe move the Mixin to src/diffusers/loaders ?

I am okay. I just followed what `transformers` did to have a unified structure. LMK.

**Comment by DN6** [comment]
> This is what I think too. So, instead of adding the PEFT mixin to `ModelMixin` we can be selective and only add to the models that we know are used in LoRA training (from our official scripts). WDYT?

I like this idea. It also keeps ModelMixin from getting bloated.

File: src/diffusers/loaders/__init__.py [hunk 2]
```diff
@@ -64,6 +64,8 @@ def text_encoder_attn_modules(text_encoder):
         _import_structure["textual_inversion"] = ["TextualInversionLoaderMixin"]
         _import_structure["ip_adapter"] = ["IPAdapterMixin"]
 
+_import_structure["peft"] = ["PeftAdapterMixin"]
```
Comment (sayakpaul) [inline src/diffusers/loaders/__init__.py:67]: This is out of the condition because:

* It's gonna be directly used in the core models (wherever applicable). 
* In our core test environment, we don't install `peft`. So, if we add the above under `is_peft_available()`, we will run into import errors. 

Can move it under `is_torch_available()`, though. 

@DN6 let me know if you have a better way of handling this.

File: src/diffusers/loaders/__init__.py [hunk 3]
```diff
@@ -76,6 +78,8 @@ def text_encoder_attn_modules(text_encoder):
             from .lora import LoraLoaderMixin, StableDiffusionXLLoraLoaderMixin
             from .single_file import FromSingleFileMixin
             from .textual_inversion import TextualInversionLoaderMixin
+
+    from .peft import PeftAdapterMixin
```
Comment (sayakpaul) [inline src/diffusers/loaders/__init__.py:82]: Same as https://github.com/huggingface/diffusers/pull/6416/files#r1440090191.

File: src/diffusers/models/modeling_utils.py [hunk 4]
```diff
@@ -183,7 +182,7 @@ def load(module: torch.nn.Module, prefix: str = ""):
     return error_msgs
 
 
-class ModelMixin(torch.nn.Module, PushToHubMixin):
+class ModelMixin(torch.nn.Module, PushToHubMixin, PeftAdapterMixin):
```
Comment (sayakpaul) [inline src/diffusers/models/modeling_utils.py:185]: Will remove from `ModelMixin` and add the separate model classes, instead, as discussed.

**Comment by sayakpaul** [comment]
For the failing test: https://github.com/huggingface/diffusers/pull/6426#issuecomment-1875396018.

**Comment by sayakpaul** [comment]
@patrickvonplaten addressed your comments:

* Moved `PeftAdapterMixin` to `src/diffusers/loaders`. 
* Added `PeftAdapterMixin` as an inheritance to only a few selected models that are used in our training scripts. 
* Added a doc. 

This is my open question: https://github.com/huggingface/diffusers/pull/6416#discussion_r1440090191

@DN6 LMK your thoughts too.

**Review by patrickvonplaten — APPROVED** [review]
Nice that works for me!

**Comment by sayakpaul** [comment]
@stevhliu feel free to edit the docs introduced in this PR in a follow-up.

**Review by younesbelkada — COMMENTED** [review]
This is very clean thanks @sayakpaul and team!
